### ISSUE TITLE
Set up SSO with mines.edu gmail accounts.

### ISSUE THREAD

**Issue by snowe2010** [ColoradoSchoolOfMines/schedule_miner#10]
We need to begin setting up single sign on through Mines' Google Apps account. The only person I know who has done this is Keith Hellman. 

I believe the best way to implement will be with Devise and OmniAuth.

**Comment by pianohacker** [comment]
I have also done this, a bit of horn-tooting which I think is pertinent. We will make it happen! And I think going into Trailhead would be difficult, as that is on a separate authentication system.

**Comment by snowe2010** [comment]
Yeah I've done it too, I'm actually almost done implementing the gem and basic routes lol. I'll push in a bit. I've also separated this into another branch of its own called OmniAuth. I found that we have no need of Devise (as far as I can tell?) because all we are doing is SSO, so no need to use users for other stuff. That way we don't really have to deal with user information, just session data :D

**Comment by ybakos** [comment]
"so no need to use users for other stuff. That way we don't really have to deal with user information,"

Exactly. This is a critical feature for university administration acceptance, if still important.

**Comment by snowe2010** [comment]
So it looks like Hellman integrated his site with the mymail.mines.edu domain using oauth2, but I don't think we have any need for that. It would be better to use OpenID, but I don't know if OpenID is enabled on Mines' apps account. Does anyone happen to know if it is and who is in charge of the domain?

**Comment by snowe2010** [comment]
Looks like it's not enabled for the mines domain. Shoot, well should we go with OAuth2 or try and get the administration to enable SSO?
![sso_minesedu](https://f.cloud.github.com/assets/1534535/1442796/5bfa1fae-41c9-11e3-9037-ab2a1964ac7d.PNG)

**Comment by sgonzalez** [comment]
Using OAuth2 would be much quicker than getting CCIT to change something. I have experience with OAuth2 (w/ Ruby on Rails and Objective-C, though) in case anyone has any questions.

**Comment by snowe2010** [comment]
The problem is we need an API key for OAuth2 and we have to specify what kind of data we want. We really don't want any data though, we just want to **authenticate** the user to make sure they have a mines email and so that we can store session data corresponding to them. 
We have no need of authorization which is the main point of OAuth. It is explained pretty well [here](http://stackoverflow.com/questions/1087031/whats-the-difference-between-openid-and-oauth?lq=1). I just don't think it's safe stating that we need user info, when we don't. 
Also, OAuth will request access to user's data and they might choose to not log in because of that.

**Comment by thevoiceless** [comment]
If I remember correctly, when using OAuth it says exactly what information we'll access. I believe the minimum an be something like "your name", which I assume most people wouldn't have a problem with.

**Comment by snowe2010** [comment]
Ah ok, well that sounds fine then. Hopefully. I've submitted a ticket requesting both OpenID and if that isn't possible then OAuth2, because I think OpenID is more desirable, but less likely.

**Comment by heck1701** [comment]
One gotcha is that this is dependent on MyMail staying with internal Google authentication... there is some motion on bringing that authentication to campus using either CAS or Shibboleth, which would break the OAuth access.

**Comment by snowe2010** [comment]
I just made some administrative decisions so I used OAuth2. I've implemented basic user functionality and Sign In/Sign Out functionality. I don't know enough about security holes through rails yet so I don't know if my code has any, if someone could review it that would be great. 
If Mines breaks our OAuth2 functionality it will most likely be between semesters, so I think we can just use OAuth2 until that time. The current schedule miner already wipes user accounts between semesters so I don't think it will be a problem losing user accounts unless we want to implement a system like myedu.com
